src/main/java/ch/njol/skript/SkriptAddon.java

```diff
@@ -21,123 +21,195 @@
 
-			for (JarEntry e : new EnumerationIterable<>(jar.entries())) {
-				if (e.getName().startsWith(basePackage) && e.getName().endsWith(".class")) {
+		int depth = !recursive ? StringUtils.count(basePackage, '/') + 1 : 0;
+
+		File file = getFile();
+		if (file == null) {
+			Skript.error("Unable to retrieve file from addon '" + getName() + "'. Classes will not be loaded.");
+			return this;
+		}
+
+		try (JarFile jar = new JarFile(file)) {
+			List<String> classNames = new ArrayList<>();
+			boolean hasWithClass = withClass != null;
+			if (entryCache == null)
+				entryCache = jar.stream().toArray(JarEntry[]::new);
+			for (int i = 0; i < entryCache.length; i++) {
+				JarEntry e = entryCache[i];
+				if (e == null) // This entry has already been loaded before
+					continue;
+				String name = e.getName();
+				if (name.startsWith(basePackage) && name.endsWith(".class") && (recursive || StringUtils.count(name, '/') <= depth)) {
 					boolean load = subPackages.length == 0;
-					for (String sub : subPackages) {
-						if (e.getName().startsWith(sub, basePackage.length())) {
+					for (String subPackage : subPackages) {
+						if (e.getName().startsWith(subPackage, basePackage.length())) {
 							load = true;
 							break;
 						}
 					}
-
-					if (load)
+					if (load) {
 						classNames.add(e.getName().replace('/', '.').substring(0, e.getName().length() - ".class".length()));
+						entryCache[i] = null; // Remove this item from the entry cache as this method will only load it once
+					}
 				}
 			}
-
 			classNames.sort(String::compareToIgnoreCase);
-
 			for (String c : classNames) {
 				try {
-					Class.forName(c, true, plugin.getClass().getClassLoader());
+					Class<?> clazz = Class.forName(c, initialize, plugin.getClass().getClassLoader());
+					if (hasWithClass)
+						withClass.accept(clazz);
 				} catch (ClassNotFoundException ex) {
-					Skript.exception(ex, "Cannot load class " + c + " from " + this);
+					Skript.exception(ex, "Cannot load class " + c);
 				} catch (ExceptionInInitializerError err) {
 					Skript.exception(err.getCause(), this + "'s class " + c + " generated an exception while loading");
 				}
 			}
-		} finally {
-			try {
-				jar.close();
-			} catch (IOException e) {}
+		} catch (IOException e) {
+			Skript.exception(e, "Failed to load classes for addon: " + plugin.getName());
 		}
 		return this;
 	}
+
+	/**
+	 * Loads all module classes found in the package search.
+	 * @param basePackage The base package to start searching in (e.g. 'ch.njol.skript').
+	 * @param subPackages Specific subpackages to search in (e.g. 'conditions').
+	 *                    If no subpackages are provided, all subpackages will be searched.
+	 *                    Note that the search will go no further than the first layer of subpackages.
+	 *                    Note that this method will also clear the entry cache of ALL checked classes,
+	 *                    	even those that are not actually a Module.
```

Comment:
https://github.com/SkriptLang/Skript/pull/4573/files/d4e01a4cad1b8120a7b16de0762b4a1ca02e52c1#diff-05cff75b740dd11508692da8390fc5de08fb75a16bcff01c63f78a21204e1733R100-R105

Public access is mainly for "just in case" scenarios. I doubt there are any addons loading the same class multiple times though, so we could probably internalize this if you think it's best :)